README.md
# Bot for Prometheus' Alertmanager


[![Docker Pulls](https://img.shields.io/docker/pulls/metalmatze/alertmanager-bot.svg?maxAge=604800)](https://hub.docker.com/r/metalmatze/alertmanager-bot)
[![Go Report Card](https://goreportcard.com/badge/github.com/nekomeowww/alertmanager-bot)](https://goreportcard.com/report/github.com/nekomeowww/alertmanager-bot)


This is the [Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) bot for
[Prometheus](https://prometheus.io/) that notifies you on alerts.  
Just configure the Alertmanager to send Webhooks to the bot and that's it.

Additionally you can always **send commands** to get up-to-date information from the alertmanager.

### Why?

Alertmanager already integrates a lot of different messengers as receivers for alerts.  
I want to extend this basic functionality.

Previously the Alertmanager could only talk to you via a chat, but now you can talk back via [commands](#commands).  
You can ask about current ongoing [alerts](#alerts) and [silences](#silences).  
In the future I plan to also support silencing via the chat, so you can silences after getting an alert from within the chat.  
A lot of other things can be added!

## Messengers

Right now it supports [Telegram](https://telegram.org/), but I'd like to [add more](#more-messengers) in the future.

## Commands

###### /start

> Hey, Matthias! I will now keep you up to date!  
> [/help](#help)

###### /stop

> Alright, Matthias! I won't talk to you again.  
> [/help](#help)

###### /alerts

> 🔥 **FIRING** 🔥  
> **NodeDown** (Node scraper.krautreporter:8080 down)  
> scraper.krautreporter:8080 has been down for more than 1 minute.  
> **Started**: 1 week 2 days 3 hours 50 minutes 42 seconds ago  
>
> 🔥 **FIRING** 🔥
> **monitored_service_down** (MONITORED SERVICE DOWN)
> The monitoring service 'digitalocean-exporter' is down.
> **Started**: 10 seconds ago

###### /silences

> NodeDown 🔕  
>  `job="ranch-eye" monitor="exporter-metrics" severity="page"`  
> **Started**: 1 month 1 week 5 days 8 hours 27 minutes 57 seconds ago  
> **Ends**: -11 months 2 weeks 2 days 19 hours 15 minutes 24 seconds  
>
> RancherServiceState 🔕  
>  `job="rancher" monitor="exporter-metrics" name="scraper" rancherURL="http://rancher.example.com/v1" severity="page" state="inactive"`  
> **Started**: 1 week 2 days 3 hours 46 minutes 21 seconds ago  
> **Ends**: -3 weeks 1 day 13 minutes 24 seconds  

###### /chats

> Currently these chat have subscribed:
> @MetalMatze


###### /status

> **AlertManager**  
> Version: 0.5.1  
> Uptime: 3 weeks 1 day 6 hours 15 minutes 2 seconds  
> **AlertManager Bot**  
> Version: 0.4.3  
> Uptime: 3 weeks 1 hour 17 minutes 19 seconds  

###### /help

> I'm a Prometheus AlertManager Bot for Telegram. I will notify you about alerts.  
> You can also ask me about my [/status](#status), [/alerts](#alerts) & [/silences](#silences)  
>   
> Available commands:  
> [/start](#start) - Subscribe for alerts.  
> [/stop](#stop) - Unsubscribe for alerts.  
> [/status](#status) - Print the current status.  
> [/alerts](#alerts) - List all alerts.  
> [/silences](#silences) - List all silences.  
> [/chats](#chats) - List all users and group chats that subscribed.

## Installation

### Docker

`docker pull metalmatze/alertmanager-bot:0.4.3`

Start as a command:

#### Bolt Storage

```bash
docker run -d \
	-e 'ALERTMANAGER_URL=http://alertmanager:9093' \
	-e 'BOLT_PATH=/data/bot.db' \
	-e 'STORE=bolt' \
	-e 'TELEGRAM_ADMIN=1234567' \
	-e 'TELEGRAM_TOKEN=XXX' \
	-v '/srv/monitoring/alertmanager-bot:/data' \
	--name alertmanager-bot \
	metalmatze/alertmanager-bot:0.4.3
```

#### Consul Storage

```bash
docker run -d \
	-e 'ALERTMANAGER_URL=http://alertmanager:9093' \
	-e 'CONSUL_URL=localhost:8500' \
	-e 'STORE=consul' \
	-e 'TELEGRAM_ADMIN=1234567' \
	-e 'TELEGRAM_TOKEN=XXX' \
	--name alertmanager-bot \
	metalmatze/alertmanager-bot:0.4.3
```

#### ETCD Storage

```bash
docker run -d \
	-e 'ALERTMANAGER_URL=http://alertmanager:9093' \
	-e 'ETCD_URL=localhost:2379' \
	-e 'STORE=etcd' \
	-e 'ETCD_TLS_INSECURE=true' \
	-e 'TELEGRAM_ADMIN=1234567' \
	-e 'TELEGRAM_TOKEN=XXX' \
	--name alertmanager-bot \
	metalmatze/alertmanager-bot:0.4.2
```

### docker-compose:

[embedmd]:# (deployments/examples/docker-compose.yaml)
```yaml
networks:
  alertmanager-bot: {}
services:
  alertmanager-bot:
    command:
    - --alertmanager.url=http://localhost:9093
    - --log.level=info
    - --store=bolt
    - --bolt.path=/data/bot.db
    environment:
      TELEGRAM_ADMIN: "1234"
      TELEGRAM_TOKEN: XXXXXXX
    image: metalmatze/alertmanager-bot:0.4.3
    networks:
    - alertmanager-bot
    ports:
    - 8080:8080
    restart: always
    volumes:
    - ./data:/data
version: "3"
```

### Kubernetes:

[embedmd]:# (deployments/examples/kubernetes.yaml)
```yaml
apiVersion: v1
items:
- apiVersion: v1
  data:
    admin: MTIzNA==
    token: WFhYWFhYWA==
  kind: Secret
  metadata:
    labels:
      app.kubernetes.io/name: alertmanager-bot
    name: alertmanager-bot
    namespace: monitoring
  type: Opaque
- apiVersion: v1
  kind: Service
  metadata:
    labels:
      app.kubernetes.io/name: alertmanager-bot
    name: alertmanager-bot
    namespace: monitoring
  spec:
    ports:
    - name: http
      port: 8080
      targetPort: 8080
    selector:
      app.kubernetes.io/name: alertmanager-bot
- apiVersion: apps/v1
  kind: StatefulSet
  metadata:
    labels:
      app.kubernetes.io/name: alertmanager-bot
    name: alertmanager-bot
    namespace: monitoring
  spec:
    podManagementPolicy: OrderedReady
    replicas: 1
    selector:
      matchLabels:
        app.kubernetes.io/name: alertmanager-bot
    serviceName: alertmanager-bot
    template:
      metadata:
        labels:
          app.kubernetes.io/name: alertmanager-bot
        name: alertmanager-bot
        namespace: monitoring
      spec:
        containers:
        - args:
          - --alertmanager.url=http://localhost:9093
          - --log.level=info
          - --store=bolt
          - --bolt.path=/data/bot.db
          env:
          - name: TELEGRAM_ADMIN
            valueFrom:
              secretKeyRef:
                key: admin
                name: alertmanager-bot
          - name: TELEGRAM_TOKEN
            valueFrom:
              secretKeyRef:
                key: token
                name: alertmanager-bot
          image: metalmatze/alertmanager-bot:0.4.3
          imagePullPolicy: IfNotPresent
          name: alertmanager-bot
          ports:
          - containerPort: 8080
            name: http
          resources:
            limits:
              cpu: 100m
              memory: 128Mi
            requests:
              cpu: 25m
              memory: 64Mi
          volumeMounts:
          - mountPath: /data
            name: data
        restartPolicy: Always
        volumes:
        - name: data
          persistentVolumeClaim:
            claimName: data
    volumeClaimTemplates:
    - apiVersion: v1
      kind: PersistentVolumeClaim
      metadata:
        labels:
          app.kubernetes.io/name: alertmanager-bot
        name: alertmanager-bot
        namespace: monitoring
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1Gi
        storageClassName: standard
kind: List
```

### Ansible

If you prefer using configuration management systems (like Ansible) you might be interested in the following role:  [mbaran0v.alertmanager-bot](https://github.com/mbaran0v/ansible-role-alertmanager-bot)

### Build from source

`GO111MODULE=on go get github.com/nekomeowww/alertmanager-bot/cmd/alertmanager-bot`

### Configuration

| ENV Variable                  | CLI flag                    | Required | Default                 | Description                                                                                                                                                                                                                          |   |   |   |
|-------------------------------|-----------------------------|----------|-------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---|---|---|
| ALERTMANAGER_URL              | alertmanager.url            |          | http://localhost:9093   | Address of the alertmanager                                                                                                                                                                                                          |   |   |   |
| BOLT_PATH                     | bolt.path                   |          | /tmp/bot.db             | Path on disk to the file where the boltdb is stored                                                                                                                                                                                  |   |   |   |
| CONSUL_URL                    | consul.url                  |          | localhost:8500          | The URL to use to connect with Consul                                                                                                                                                                                                |   |   |   |
| LISTEN_ADDR                   | listen.addr                 |          | 0.0.0.0:8080            | Address that the bot listens for webhooks                                                                                                                                                                                            |   |   |   |
| STORE                         | store                       | ✓        |                         | The type of the store to use, choose from bolt (local), consul or etcd (distributed)                                                                                                                                                 |   |   |   |
| STORE_KEY_PREFIX              | storeKeyPrefix              |          | telegram/chats          | Key prefix for the store                                                                                                                                                                                                             |   |   |   |
| ETCD_URL                      | etcd.url                    |          | localhost:2379          | The URL that's used to connect to the ETCD store                                                                                                                                                                                     |   |   |   |
| ETCD_TLS_INSECURE             | etcd.tls.insecure           |          | false                   | Use TLS connection to ETCD store or not                                                                                                                                                                                              |   |   |   |
| ETCD_TLS_INSECURE_SKIP_VERIFY | etcd.tls.insecureSkipVerify |          |                         | Skip server certificates verification                                                                                                                                                                                                |   |   |   |
| ETCD_TLS_CERT                 | etcd.tls.cert               |          |                         | Path to the TLS cert file                                                                                                                                                                                                            |   |   |   |
| ETCD_TLS_KEY                  | etcd.tls.key                |          |                         | Path to the TLS key file                                                                                                                                                                                                             |   |   |   |
| ETCD_TLS_CACERT               | etcd.tls.ca                 |          |                         | Path to the TLS trusted CA cert file                                                                                                                                                                                                 |   |   |   |
| LOG_JSON                      | log.json                    |          |                         | Tell the application to log json and not key value pairs                                                                                                                                                                             |   |   |   |
| LOG_LEVEL                     | log.level                   |          | info                    | The log level to use for filtering logs. Possible values: debug, info, warn, error                                                                                                                                                   |   |   |   |
| TELEGRAM_ADMIN                | telegram.admin              | ✓        |                         | The Telegram user id for the admin (not the bot itself, you, the user). The bot will only reply to messages sent from an admin. All other messages are dropped and logged on the bot's console.  Your user id you can get from [@userinfobot](https://t.me/userinfobot). |   |   |   |
| TELEGRAM_TOKEN                | telegram.token              | ✓        |                         | Token you get from [@botfather](https://telegram.me/botfather)                                                                                                                                                                       |   |   |   |
| TEMPLATE_PATHS                | template.paths              |          | /templates/default.tmpl | Path to custom message templates                                                                                                                                                                                                     |   |   |   |

#### Authentication

Additional users may be allowed to command the bot by giving multiple instances
of the `--telegram.admin` command line option or by specifying a
newline-separated list of telegram user IDs in the `TELEGRAM_ADMIN` environment
variable.
```
- TELEGRAM_ADMIN="**********\n************"
--telegram.admin=1 --telegram.admin=2
```
#### Alertmanager Configuration

Now you need to connect the Alertmanager to send alerts to the bot.  
A webhook is used for that, so make sure your `LISTEN_ADDR` is reachable for the Alertmanager.

For example add this to your `alertmanager.yml` configuration:
```yaml
receivers:
- name: 'alertmanager-bot'
  webhook_configs:
  - send_resolved: true
    url: 'http://alertmanager-bot:8080'
```

## Development

Build the binary using `make`:

```
make
```

In case you have `$GOPATH/bin` in your `$PATH` you can now simply start the bot by running:

```bash
alertmanager-bot
```

## Missing

##### Commands

* `/silence` - show a specific silence  
* `/silence_del` - delete a silence by command  
* `/silence_add` - add a silence for a alert by command

##### More Messengers

At the moment I only implemented Telegram, because it's so freakin' easy to do.

Messengers considered to add in the future:

* [Slack](https://slack.com/)
* [Mattermost](https://about.mattermost.com/)
* [Matrix](https://matrix.org/)

If one is missing for you just open an issue.


CHANGELOG.md
## 0.4.3 / 2021-01-11

* [ENHANCEMENT] Support description annotation which is used by newer alerts [#113](https://github.com/nekomeowww/alertmanager-bot/pull/113).
* [ENHANCEMENT] Create jsonnet-based deployments for Kubernets and Docker Compose [#115](https://github.com/nekomeowww/alertmanager-bot/pull/115).
* [BUGFIX] Don't have a fix "EXPOSE" as the port is configurable [#106](https://github.com/nekomeowww/alertmanager-bot/pull/106).

## 0.4.2 / 2020-01-11

* [BUGFIX] Fix flags with defaults that aren't required anymore [#96](https://github.com/nekomeowww/alertmanager-bot/pull/96).

## 0.4.1 / 2020-01-07

Update to Go 1.13 and dependencies.

* [BUGFIX] Fix default template path [#55](https://github.com/nekomeowww/alertmanager-bot/pull/55).
* [BUGFIX] Bring back defaults for flags [#93](https://github.com/nekomeowww/alertmanager-bot/pull/93).

## 0.4.0 / 2019-02-19

* [FEATURE] Add ability to use templates for Telegram messages [#32](https://github.com/nekomeowww/alertmanager-bot/pull/32).
* [ENHANCEMENT] Truncate too large messages [#52](https://github.com/nekomeowww/alertmanager-bot/pull/52), thanks @BulatSaif.

## 0.3.1 / 2018-06-11

* [BUGFIX] Escape emojis in messages [#22](https://github.com/nekomeowww/alertmanager-bot/pull/22), thanks @caarlos0.

## 0.3.0 / 2018-05-15

* [FEATURE] Allow for multiple bot admin users [#19](https://github.com/nekomeowww/alertmanager-bot/pull/19), thanks @slrz.
* [FEATURE] Add log.level and log.json as cli flags [[9050ff4]](https://github.com/nekomeowww/alertmanager-bot/commit/9050ff418bf5a07fcd684fb01fa7838a36b0af38).
* [ENHANCEMENT] Handle Ctrl+C interrupts and shutdown bot gracefully [[07c2356]](https://github.com/nekomeowww/alertmanager-bot/commit/07c23563800e62e97cc0437a47cefd1aea332a82).
* [ENHANCEMENT] Internal refactoring of HTTP calls [#15](https://github.com/nekomeowww/alertmanager-bot/pull/15), thanks @vtolstov.

## 0.2.1 / 2017-11-04

* [BUGFIX] /chats command send to the sender and not the chat [#10](https://github.com/nekomeowww/alertmanager-bot/issues/10)

## 0.2.0 / 2017-10-17

* [BREAKING] Change the `STORE` env var to switch on the store backend instead of path
* [FEATURE] The Bot now ignores its own name as suffix on command in a group chat [[aae2dc5]](https://github.com/nekomeowww/alertmanager-bot/commit/aae2dc5c1dae5f865cd697cb649fb757b7efaa6f) 
* [FEATURE] Use libkv as store backend [[6419064]](https://github.com/nekomeowww/alertmanager-bot/commit/64190646e71910a10fdcb6e7533dc8a8dc485fec)
* [ENHANCEMENT] Use [dep](https://github.com/golang/dep) to pin dependencies [[b7e085b]](https://github.com/nekomeowww/alertmanager-bot/commit/4bfd3f7a2ec559eee712f37ba8e32ca848717905)
* [ENHANCEMENT] Use the Option pattern to change the Bot's default values in `NewBot()`.

## 0.1.0 / 2017-03-31

Initial release.


tests/workflows/telegram/id_test.go
package telegram

import (
	"github.com/nekomeowww/alertmanager-bot/pkg/telegram"
	"gopkg.in/tucnak/telebot.v2"
)

var idWorkflows = []workflow{{
	name: "IDAsNobody",
	messages: []telebot.Update{{
		Message: &telebot.Message{
			Sender: nobody,
			Chat:   chatFromUser(nobody),
			Text:   telegram.CommandID,
		},
	}},
	replies: []reply{{
		recipient: "222",
		message:   "Your ID is 222",
	}},
	counter: map[string]uint{telegram.CommandID: 1},
	logs: []string{
		"level=debug msg=\"message received\" text=/id",
	},
}, {
	name: "IDAsAdmin",
	messages: []telebot.Update{{
		Message: &telebot.Message{
			Sender: admin,
			Chat:   chatFromUser(admin),
			Text:   telegram.CommandID,
		},
	}},
	replies: []reply{{
		recipient: "123",
		message:   "Your ID is 123",
	}},
	counter: map[string]uint{telegram.CommandID: 1},
	logs: []string{
		"level=debug msg=\"message received\" text=/id",
	},
}}


tests/workflows/telegram/help_test.go
package telegram

import (
	"strings"

	"github.com/nekomeowww/alertmanager-bot/pkg/telegram"
	"gopkg.in/tucnak/telebot.v2"
)

var helpWorkflows = []workflow{{
	name: "Help",
	messages: []telebot.Update{{
		Message: &telebot.Message{
			Sender: admin,
			Chat:   chatFromUser(admin),
			Text:   telegram.CommandHelp,
		},
	}},
	replies: []reply{{
		recipient: "123",
		message:   strings.TrimSpace(telegram.ResponseHelp),
	}},
	counter: map[string]uint{telegram.CommandHelp: 1},
	logs: []string{
		"level=debug msg=\"message received\" text=/help",
	},
}, {
	name: "HelpAsNobody",
	messages: []telebot.Update{{
		Message: &telebot.Message{
			Sender: nobody,
			Chat:   chatFromUser(nobody),
			Text:   telegram.CommandHelp,
		},
	}},
	replies: []reply{},
	logs: []string{
		"level=info msg=\"dropping message from forbidden sender\" sender_id=222 sender_username=nobody",
	},
}}


tests/workflows/telegram/stop_test.go
package telegram

import (
	"github.com/nekomeowww/alertmanager-bot/pkg/telegram"
	"gopkg.in/tucnak/telebot.v2"
)

var stopWorkflows = []workflow{{
	name: "StopWithoutStart",
	messages: []telebot.Update{{
		Message: &telebot.Message{
			Sender: admin,
			Chat:   chatFromUser(admin),
			Text:   telegram.CommandStop,
		},
	}},
	replies: []reply{{
		recipient: "123",
		message:   "Alright, Elliot! I won't talk to you again.\n/help",
	}},
	counter: map[string]uint{telegram.CommandStop: 1},
	logs: []string{
		"level=debug msg=\"message received\" text=/stop",
		"level=info msg=\"user unsubscribed\" username=elliot user_id=123",
	},
}}


cmd/alertmanager-bot/main.go
package main

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"runtime"
	"strings"
	"syscall"
	"time"

	"github.com/alecthomas/kong"
	"github.com/docker/libkv/store"
	"github.com/docker/libkv/store/boltdb"
	"github.com/docker/libkv/store/consul"
	"github.com/docker/libkv/store/etcd"
	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/log/level"
	"github.com/nekomeowww/alertmanager-bot/pkg/alertmanager"
	"github.com/nekomeowww/alertmanager-bot/pkg/telegram"
	"github.com/oklog/run"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
)

const (
	storeBolt   = "bolt"
	storeConsul = "consul"
	storeEtcd   = "etcd"

	levelDebug = "debug"
	levelInfo  = "info"
	levelWarn  = "warn"
	levelError = "error"
)

var (
	// Version of alertmanager-bot.
	Version string
	// Revision or Commit this binary was built from.
	Revision string
	// GoVersion running this binary.
	GoVersion = runtime.Version()
	// StartTime has the time this was started.
	StartTime = time.Now()
)

var cli struct {
	AlertmanagerURL *url.URL `name:"alertmanager.url" default:"http://localhost:9093/" help:"The URL that's used to connect to the alertmanager"`
	ListenAddr      string   `name:"listen.addr" default:"0.0.0.0:8080" help:"The address the alertmanager-bot listens on for incoming webhooks"`
	LogJSON         bool     `name:"log.json" default:"false" help:"Tell the application to log json and not key value pairs"`
	LogLevel        string   `name:"log.level" default:"info" enum:"error,warn,info,debug" help:"The log level to use for filtering logs"`
	TemplatePaths   []string `name:"template.paths" default:"/templates/default.tmpl" help:"The paths to the template"`

	cliTelegram

	Store       string `required:"true" name:"store" enum:"bolt,consul,etcd" help:"The store to use"`
	StorePrefix string `name:"storeKeyPrefix" default:"telegram/chats" help:"Prefix for store keys"`
	cliBolt
	cliConsul
	cliEtcd
}

type cliBolt struct {
	Path string `name:"bolt.path" type:"path" default:"/tmp/bot.db" help:"The path to the file where bolt persists its data"`
}

type cliConsul struct {
	URL *url.URL `name:"consul.url" default:"localhost:8500" help:"The URL that's used to connect to the consul store"`
}

type cliEtcd struct {
	URL                   *url.URL `name:"etcd.url" default:"localhost:2379" help:"The URL that's used to connect to the etcd store"`
	TLSInsecure           bool     `name:"etcd.tls.insecure" default:"false" help:"Use TLS or not"`
	TLSInsecureSkipVerify bool     `name:"etcd.tls.insecureSkipVerify" default:"false" help:"Skip server certificates verification"`
	TLSCert               string   `name:"etcd.tls.cert" type:"path" help:"Path to the TLS cert file"`
	TLSKey                string   `name:"etcd.tls.key" type:"path" help:"Path to the TLS key file"`
	TLSCA                 string   `name:"etcd.tls.ca" type:"path" help:"Path to the TLS trusted CA cert file"`
}

type cliTelegram struct {
	Admins []int  `required:"true" name:"telegram.admin" help:"The ID of the initial Telegram Admin"`
	Token  string `required:"true" name:"telegram.token" env:"TELEGRAM_TOKEN" help:"The token used to connect with Telegram"`
}

func main() {
	_ = kong.Parse(&cli,
		kong.Name("alertmanager-bot"),
	)

	var err error

	levelFilter := map[string]level.Option{
		levelError: level.AllowError(),
		levelWarn:  level.AllowWarn(),
		levelInfo:  level.AllowInfo(),
		levelDebug: level.AllowDebug(),
	}

	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
	if cli.LogJSON {
		logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
	}

	logger = level.NewFilter(logger, levelFilter[cli.LogLevel])
	logger = log.With(logger,
		"ts", log.DefaultTimestampUTC,
		"caller", log.DefaultCaller,
	)

	reg := prometheus.NewRegistry()
	reg.MustRegister(
		prometheus.NewGoCollector(),
		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
	)

	var am *alertmanager.Client
	{
		client, err := alertmanager.NewClient(cli.AlertmanagerURL)
		if err != nil {
			level.Error(logger).Log("msg", "failed to create alertmanager client", "err", err)
			os.Exit(1)
		}
		am = client
	}

	var kvStore store.Store
	{
		switch strings.ToLower(cli.Store) {
		case storeBolt:
			kvStore, err = boltdb.New([]string{cli.cliBolt.Path}, &store.Config{Bucket: "alertmanager"})
			if err != nil {
				level.Error(logger).Log("msg", "failed to create bolt store backend", "err", err)
				os.Exit(1)
			}
		case storeConsul:
			kvStore, err = consul.New([]string{cli.cliConsul.URL.String()}, nil)
			if err != nil {
				level.Error(logger).Log("msg", "failed to create consul store backend", "err", err)
				os.Exit(1)
			}
		case storeEtcd:
			tlsConfig := &tls.Config{}

			if cli.cliEtcd.TLSCert != "" {
				cert, err := tls.LoadX509KeyPair(cli.cliEtcd.TLSCert, cli.cliEtcd.TLSKey)
				if err != nil {
					level.Error(logger).Log("msg", "failed to create etcd store backend, could not load certificates", "err", err)
					os.Exit(1)
				}
				tlsConfig.Certificates = []tls.Certificate{cert}
			}

			if cli.cliEtcd.TLSCA != "" {
				caCert, err := ioutil.ReadFile(cli.cliEtcd.TLSCA)
				if err != nil {
					level.Error(logger).Log("msg", "failed to create etcd store backend, could not load ca certificate", "err", err)
					os.Exit(1)
				}

				caCertPool := x509.NewCertPool()
				caCertPool.AppendCertsFromPEM(caCert)
				tlsConfig.RootCAs = caCertPool
			}

			tlsConfig.InsecureSkipVerify = cli.cliEtcd.TLSInsecureSkipVerify

			if !cli.cliEtcd.TLSInsecure {
				kvStore, err = etcd.New([]string{cli.cliEtcd.URL.String()}, &store.Config{TLS: tlsConfig})
			} else {
				kvStore, err = etcd.New([]string{cli.cliEtcd.URL.String()}, nil)
			}

			if err != nil {
				level.Error(logger).Log("msg", "failed to create etcd store backend", "err", err)
				os.Exit(1)
			}
		default:
			level.Error(logger).Log("msg", "please provide one of the following supported store backends: bolt, consul, etcd")
			os.Exit(1)
		}
	}
	defer kvStore.Close()

	ctx, cancel := context.WithCancel(context.Background())

	// TODO Needs fan out for multiple bots
	webhooks := make(chan alertmanager.TelegramWebhook, 32)

	var g run.Group
	{
		tlogger := log.With(logger, "component", "telegram")

		commandCounter := prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "alertmanagerbot_commands_total",
			Help: "Number of commands received by command name",
		}, []string{"command"})
		reg.MustRegister(commandCounter)

		commandCount := func(command string) {
			commandCounter.WithLabelValues(command).Inc()
		}

		chats, err := telegram.NewChatStore(kvStore, cli.StorePrefix)
		if err != nil {
			level.Error(logger).Log("msg", "failed to create chat store", "err", err)
			os.Exit(1)
		}

		bot, err := telegram.NewBot(
			chats, cli.cliTelegram.Token, cli.cliTelegram.Admins[0],
			telegram.WithLogger(tlogger),
			telegram.WithCommandEvent(commandCount),
			telegram.WithAddr(cli.ListenAddr),
			telegram.WithAlertmanager(am),
			telegram.WithTemplates(cli.AlertmanagerURL, cli.TemplatePaths...),
			telegram.WithRevision(Revision),
			telegram.WithStartTime(StartTime),
			telegram.WithExtraAdmins(cli.cliTelegram.Admins[1:]...),
		)
		if err != nil {
			level.Error(tlogger).Log("msg", "failed to create bot", "err", err)
			os.Exit(2)
		}

		g.Add(func() error {
			level.Info(tlogger).Log(
				"msg", "starting alertmanager-bot",
				"version", Version,
				"revision", Revision,
				"goVersion", GoVersion,
			)

			// Runs the bot itself communicating with Telegram
			return bot.Run(ctx, webhooks)
		}, func(err error) {
			cancel()
		})
	}
	{
		wlogger := log.With(logger, "component", "webserver")

		// TODO: Use Heptio's healthcheck library
		handleHealth := func(w http.ResponseWriter, r *http.Request) {
			w.WriteHeader(http.StatusOK)
		}

		webhooksCounter := prometheus.NewCounter(prometheus.CounterOpts{
			Name: "alertmanagerbot_webhooks_total",
			Help: "Number of webhooks received by this bot",
		})

		reg.MustRegister(webhooksCounter)

		m := http.NewServeMux()
		m.HandleFunc("/webhooks/telegram/", alertmanager.HandleTelegramWebhook(wlogger, webhooksCounter, webhooks))
		m.Handle("/metrics", promhttp.HandlerFor(reg, promhttp.HandlerOpts{}))
		m.HandleFunc("/health", handleHealth)
		m.HandleFunc("/healthz", handleHealth)

		s := http.Server{
			Addr:    cli.ListenAddr,
			Handler: m,
		}

		g.Add(func() error {
			level.Info(wlogger).Log("msg", "starting webserver", "addr", cli.ListenAddr)
			return s.ListenAndServe()
		}, func(err error) {
			_ = s.Shutdown(context.Background())
		})
	}
	{
		sig := make(chan os.Signal)
		signal.Notify(sig, os.Interrupt, syscall.SIGTERM)

		g.Add(func() error {
			<-sig
			return nil
		}, func(err error) {
			cancel()
			close(sig)
		})
	}

	if err := g.Run(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
